### PR TITLE
Add context in legacy module translation function

### PR DESCRIPTION
**Pull request by aleeks**
| Questions     | Answers
| ------------- | -------------------------------------------------------
| Branch?       | develop
| Description?  | Add context in legacy module translation function
| Type?         | improvement
| Category?     | CO
| BC breaks?    | no
| Deprecations? | no
| Fixed ticket? | 
| How to test?  | try with a module translated in many language for example

Solve this: 
```php

public function installMenu()
{
    $tab = new Tab();

    foreach (Language::getLanguages() as $lang) {
        $tab->name[$lang['id_lang']] = $this->l('My beautiful module');
    }

    $tab->class_name = 'AdminBouh';
    $tab->id_parent = 2;
    $tab->module = $this->name;
    $tab->add();

    $tab = new Tab();

    foreach (Language::getLanguages() as $lang) {
        $tab->name[$lang['id_lang']] = $this->l('My beautiful module', false, $lang['locale']);
    }

    $tab->class_name = 'AdminBouh';
    $tab->id_parent = 2;
    $tab->module = $this->name;
    $tab->add();
}
```

menu

**the first tab is not translated, the second is.**

ping @Quetzacoalt91 

It adds translations files when needed: 
```txt
array (size=1)
  'ps_bouh' => 
    array (size=1)
      'fr' => boolean true
Test de translation

array (size=1)
  'ps_bouh' => 
    array (size=2)
      'fr' => boolean true
      'en' => boolean true
Test translation

array (size=1)
  'ps_bouh' => 
    array (size=3)
      'fr' => boolean true
      'en' => boolean true
      'es' => boolean true
Traducción de prueba

array (size=1)
  'ps_bouh' => 
    array (size=4)
      'fr' => boolean true
      'en' => boolean true
      'es' => boolean true
      'it' => boolean true
Traduzione della prova
```

### CODE REVIEW

**Review by sarjon — COMMENTED** [review]
You could avoid using else and just return false in getIsoByLocale function.

File: classes/Translate.php [hunk 1]
```diff
@@ -170,31 +173,36 @@ public static function getModuleTranslation($module, $originalString, $source, $
...
+                _PS_MODULE_DIR_.$name.'/translations/'.$iso.'.php',
                 // PrestaShop 1.4 translations
-                _PS_MODULE_DIR_.$name.'/'.$language->iso_code.'.php'
+                _PS_MODULE_DIR_.$name.'/'.$iso.'.php'
             );
             foreach ($filesByPriority as $file) {
                 if (file_exists($file)) {
                     include_once($file);
-                    $_MODULES = !empty($_MODULES) ? $_MODULES + $_MODULE : $_MODULE; //we use "+" instead of array_merge() because array merge erase existing values.
+                    $_MODULES = !empty($_MODULES) ? array_merge($_MODULES, $_MODULE) : $_MODULE;
```
Comment (aleeks) [inline classes/Translate.php:197]: array_merge used to repleace FR to ES or EN when needed (for example)

**Comment by Quetzacoalt91** [comment]
While reviewing this PR, I will check to include this feature in the tab automation code (introduced in #5922)

File: classes/Translate.php [hunk 2]
```diff
@@ -170,31 +173,36 @@ public static function getModuleTranslation($module, $originalString, $source, $
...
+        if (!is_null($locale)) {
+            $iso = Language::getIsoByLocale($locale);
+        }
 
-        if (!isset($translationsMerged[$name]) && isset(Context::getContext()->language)) {
+        if (empty($iso)) {
+            $iso = Context::getContext()->language->iso_code;
+        }
+
+        if (!isset($translationsMerged[$name][$iso]) && isset(Context::getContext()->language)) {
```
Comment (Quetzacoalt91) [inline classes/Translate.php]: Is `isset(Context::getContext()->language)` still needed?

File: classes/Language.php [hunk 3]
```diff
@@ -707,14 +707,34 @@ public static function getLangDetails($iso)
...
-            return false;
         }
+
+        return false;
+    }
+
+    /**
+     * @param string $locale
+     *
+     * @return string|false|null
```
Comment (Quetzacoalt91) [inline classes/Language.php]: Are you sure that `null` can be returned?

File: classes/Language.php [hunk 4]
```diff
@@ -707,14 +707,34 @@ public static function getLangDetails($iso)
     public static function getLocaleByIso($isoCode)
     {
         if (!Validate::isLanguageIsoCode($isoCode)) {
-            throw new Exception(sprintf('The ISO code %s is invalid'));
+            throw new Exception(sprintf('The ISO code %s is invalid', $isoCode));
```
Comment (Quetzacoalt91) [inline classes/Language.php]: Actually, if we don't plan to translate it, we can remove the sprintf() and concatenate instead.

File: classes/Language.php (hunk 3 above)
Comment (aleeks) [reply classes/Language.php]: no idea, it was here, maybe if the json file is empty in some part. no matter, i remove it

**Review by Quetzacoalt91 — DISMISSED** [review]

File: classes/Language.php [hunk 5]
```diff
@@ -700,21 +700,41 @@ public static function getLangDetails($iso)
...
+    }
+
+    /**
+     * @param string $locale
+     *
+     * @return string|false
+     *
+     * @throws Exception
+     */
+    public static function getIsoByLocale($locale)
```
Comment (fatmaBouchekoua) [inline classes/Language.php:731]: Missing function description to get a proper documentation.

File: classes/Translate.php [hunk 6]
```diff
@@ -170,31 +173,36 @@ public static function getModuleTranslation($module, $originalString, $source, $
...
-                _PS_THEME_DIR_.'modules/'.$name.'/translations/'.$language->iso_code.'.php',
-                _PS_THEME_DIR_.'modules/'.$name.'/'.$language->iso_code.'.php',
+                _PS_THEME_DIR_.'modules/'.$name.'/translations/'.$iso.'.php',
+                _PS_THEME_DIR_.'modules/'.$name.'/'.$iso.'.php',
                 // PrestaShop 1.5 translations
-                _PS_MODULE_DIR_.$name.'/translations/'.$language->iso_code.'.php',
+                _PS_MODULE_DIR_.$name.'/translations/'.$iso.'.php',
                 // PrestaShop 1.4 translations
-                _PS_MODULE_DIR_.$name.'/'.$language->iso_code.'.php'
+                _PS_MODULE_DIR_.$name.'/'.$iso.'.php'
```
Comment (fatmaBouchekoua) [inline classes/Translate.php:192]: These concatenation shouldn't be spaced?

File: classes/Language.php (hunk 5 above)
Comment (aleeks) [reply classes/Language.php:731]: done

File: classes/Translate.php (hunk 6 above)
Comment (aleeks) [reply classes/Translate.php:192]: no norm for that, in this function is like that, so..

**Review by Quetzacoalt91 — APPROVED** [review]

**Pull request merged**